libtransmission/peer-mgr.cc

```diff
@@ -475,7 +475,14 @@ class tr_swarm
         if (peer_info)
         {
             peer_info->found_at(from);
-            peer_info->set_pex_flags(flags);
+            if (from == TR_PEER_FROM_PEX || from == TR_PEER_FROM_RESUME)
+            {
+                peer_info->set_pex_flags(flags);
```

Comment:
what is the logic for clearing prior flags for these two sources but not the others?

Reply:
Frankly, I'm not so sure about this implementation either.

The initial motivation for making this change is that it didn't feel right to me that the pex flags are never unset. For example, a peer might decide to disable metadata transfer (BEP-9), and we didn't have any mechanisms in place to unsetMy thinking was these 2 source supposedly would provide the complete set of pex flags, so if we got them, we should overwrite the flags we currently have.

Now I'm starting to think that this isn't a good idea, since we can't be sure how up-to-date/correct the pex flags from pex and resume are. The most trustworthy source of information is from the peers themselves, e.g. by ltep handshake.

Reply:
38b71af 84b2bb7 935ca44

I thought through the problem some more and made some refactors, see what you think.